### PR TITLE
For E0223, suggest associated functions that are similar to the path

### LINKED ISSUE #109195: confusing output when using `String::from::utf8` instead of `String::from_utf8`
### Code

```Rust
`
    pub(crate) fn to_disk(&self) {
        let original_path = self.file.clone();
        let original_path_display = original_path.display();
        let original_code = std::fs::read_to_string(&original_path).unwrap_or("<error>".into());

        let executable = self.executable.clone();
        let executable_bin = self.executable.path();

        let output: String = if let Ok(output) = std::process::Command::new(&executable_bin)
            .arg("--version")
            .arg("--verbose")
            .output()
        {
            String::from::utf8(output.stdout).unwrap()
        } else if let Ok(output_verbose) = std::process::Command::new(&executable_bin)
            .arg("--version")
            .output()
        {
            String::from::utf8(output_verbose.stdout).unwrap()
        } else {
            "<failed to get version>".to_string()
        };

        let test = format!(
            "
        File: {original_path_display}
        `
        {original_code}
        ````
        
        
        "
        );
    }

````
```


### Current output

```Shell
error[E0223]: ambiguous associated type
   --> src/ice.rs:125:13
    |
125 |             String::from::utf8(output.stdout).unwrap()
    |             ^^^^^^^^^^^^
    |
help: if there were a trait named `Example` with associated type `from` implemented for `std::string::String`, you could use the fully-qualified path
    |
125 |             <std::string::String as Example>::from::utf8(output.stdout).unwrap()
    |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

error[E0223]: ambiguous associated type
   --> src/ice.rs:130:13
    |
130 |             String::from::utf8(output_verbose.stdout).unwrap()
    |             ^^^^^^^^^^^^
    |
help: if there were a trait named `Example` with associated type `from` implemented for `std::string::String`, you could use the fully-qualified path
    |
130 |             <std::string::String as Example>::from::utf8(output_verbose.stdout).unwrap()
    |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```


### Desired output

```Shell
Item `String::from::utf8` not found, did you mean similar similar function `String::from_utf8` which also happens to return the desired target type?
```


### Rationale and extra context

This had me *VERY* confused as I was certain I would try to get a String here, where does Example come from?? And associated types wut??

The problem is I used `String::from::utf8` which should have been `String::from_utf8` :D

Maybe rustc could improve its suggestion somehow in that case?

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by trevyn**
e.g. for `String::from::utf8`, suggest `String::from_utf8`

Closes #109195

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 1]
```diff
@@ -1373,6 +1374,57 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                                && let Ok([inherent_impl]) = tcx.inherent_impls(ty_def_id)
+                                && let Some(ty::AssocItem { kind: ty::AssocKind::Fn, .. }) = tcx
+                                    .associated_items(inherent_impl)
+                                    .filter_by_name_unhygienic(Symbol::intern(&name))
+                                    .next()
+                            {
+                                true
+                            } else {
+                                qself_ty.is_str()
+                                    && ["from_utf8", "from_utf8_mut"].contains(&name.as_str())
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:1407]: imo this is too special cased for us to include here. Also, why is this not achieved with the above logic?

File: compiler/rustc_hir_analysis/src/astconv/mod.rs [hunk 2]
```diff
@@ -1373,6 +1374,57 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+                            ..
+                        }) = node
+                        {
+                            let name = format!("{ident2}_{ident3}");
+                            if if let Some(ty_def_id) = qself_ty.ty_def_id()
+                                && let Ok([inherent_impl]) = tcx.inherent_impls(ty_def_id)
+                                && let Some(ty::AssocItem { kind: ty::AssocKind::Fn, .. }) = tcx
+                                    .associated_items(inherent_impl)
+                                    .filter_by_name_unhygienic(Symbol::intern(&name))
+                                    .next()
```
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/astconv/mod.rs:1402]: Maybe use something like the [method resolution logic](https://github.com/rust-lang/rust/blob/master/compiler/rustc_hir_typeck/src/callee.rs#L87) to find the method, even if it's hidden behind derefs.

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_hir_analysis/src/astconv/mod.rs (hunk 2 above)
Comment (trevyn) [reply compiler/rustc_hir_analysis/src/astconv/mod.rs]: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/Method.20resolution.20logic.20from.20.60AstConv.60

**Comment by trevyn** [comment]
@oli-obk I moved everything into a helper function and removed the `str` special-casing. I don't quite follow how to get to method lookup from `rustc_hir_analysis`/`AstConv`, a pointer would be helpful if this still makes sense to you.

**Comment by oli-obk** [comment]
Oh... yea, there's no good way to invoke that from here :/ I guess we'll miss out on some cases, but get the usual ones.

@bors r+ rollup

**Pull request merged**